apps/src/templates/rubrics/LearningGoals.jsx

```diff
@@ -28,6 +29,149 @@ import ProgressRing from './ProgressRing';
+ */
+export function clearAnnotations() {
+  studioApp().clearAnnotations();
+  studioApp().clearHighlightedLines();
+}
+
+/**
+ * Retrieves the currently viewed source code.
+ *
+ * All comments are replaced with whitespace of equal length.
+ */
+export function getAnonymizedCode() {
+  // Find code snippet by looking at the student code
+  let code = studioApp().getCode();
+
+  // Replace comments with whitespace. Our AI does not see the comments.
+  // So, we must replace them with whitespace so we can find the right
+  // code to reference.
+
+  // This regex pattern captures three groups:
+  // 1) Single or double quoted strings
+  // 2) Multi-line comments
+  // 3) Single-line comments
+  const pattern = /(".*?[^\\]"|\'.*?[^\\]\'|\/\*.*?\*\/|\/\/[^\n]*)/gs;
+  for (const submatch of code.matchAll(pattern)) {
+    const context = submatch[0].trim();
+    if (!context.startsWith('"') && !context.startsWith("'")) {
+      code = code.replace(context, context.replace(/[^\n]/g, ' '));
+    }
+  }
+
+  return code;
+}
+
+/**
+ * Returns the first and last line that contains the given code snippet.
+ */
+export function findCodeRegion(code, lines, snippet) {
+  // Attempt to just find the code in the full code listing
+  let index = code.indexOf(snippet);
+  let lastIndex = index + snippet.length;
+  if (index < 0) {
+    // Failing to find it, we need to find it line by line instead
+    let context = snippet;
+    let position = 0;
+    for (let line of lines) {
+      // Remember the original length of the line
+      let lineLength = line.length;
+      line = line.trim();
+      if (line !== '') {
+        if (context.startsWith(line)) {
+          if (index < 0) {
+            // Remember the first position in the original code that
+            // the AI references.
+            index = position;
+          }
+          // Remember the last position in the original code.
+          lastIndex = position + lineLength;
+          context = context.substring(line.length + 1).trim();
+          if (context === '') {
+            // All of the code was found
+            break;
+          }
+        } else {
+          // We didn't match it. Reset our search.
+          context = snippet;
+          index = -1;
+        }
+      }
+
+      // Move the current search position
+      position += lineLength + 1;
+    }
+  }
+
+  // As long as we found a region, we will determine what the line
+  // number was for the first and last line of the region.
+  if (index < 0) {
+    return [null, null];
+  }
+
+  let lineNumber = (code.substring(0, index).match(/\n/g) || []).length + 1;
+  let lastLineNumber =
+    (code.substring(0, lastIndex).match(/\n/g) || []).length + 1;
+
+  return [lineNumber, lastLineNumber];
+}
+
+/**
+ * Adds annotations to the source code viewed in the current editor based
+ * on the AI observations of the learning goal referenced by the given index.
+ */
+export function annotateLines(observations) {
```

Comment:
I think it would be helpful to describe the contract in the method comments here, including what happens if the string does not match the required format. not necessarily a full breakdown of the regex, but an example of a valid input would be helpful.

Reply:
Added a whole lotta context to the docstring.

Reply:
Looks great! 😍